capstone/src/main/java/com/google/sps/data/AndroidManifestParser.java

```diff
@@ -162,13 +159,14 @@ public int LEW(byte[] arr, int off) {
 
     //This function takes in the attribute value which contains attributes and prints the
     //attributes with permission
-    public void getPermissions(String param){
+    public ArrayList<String> getPermissions(String param){
         //Uses the key to search through the attribute value and checks if it's present
         String key = "android.permission.";
         if(param.contains(key)){
             param = param.replaceAll(key, "");
             permissionsList.add(param);
         }
+    return permissionsList;
```

Comment:
For testability and readability, I would suggest created an ArrayList in this function, add all the permissions to it, then return that fresh ArrayList. The caller above would then use .addAll() to add everything to the other array list. Another option is to pass in an ArrayList to be modified by the function. It mostly helps with expectations of the user of the library who may call "getPermissions" with some random string not expecting the call to modify internal state of the parser.